sdks/python/apache_beam/transforms/environments.py

```diff
@@ -109,12 +113,25 @@ class Environment(object):
   _urn_to_env_cls = {}  # type: Dict[str, type]
 
   def __init__(self,
-               capabilities,  # type: Iterable[str]
-               artifacts,  # type: Iterable[beam_runner_api_pb2.ArtifactInformation]
-              ):
+      capabilities,  # type: Iterable[str]
+      artifacts,  # type: Iterable[beam_runner_api_pb2.ArtifactInformation]
+      resource_hints,  # type: Optional[Mapping[str, bytes]]
+               ):
     # type: (...) -> None
     self._capabilities = capabilities
     self._artifacts = artifacts
+    self._resource_hints = dict(resource_hints) if resource_hints else {}
+
+  def __eq__(self, other):
+    return (
+        self.__class__ == other.__class__
```

Comment:
I don't think this is safe, one can have multiple, distinct docker environments (for example).

Reply:
Wouldn't Docker environments be compared via `__eq__` defined on DockerEnvironment subclass that takes the image into account?

Reply:
The danger here is that a subclass forgets to override `__eq__` and this gets uses rather than, say, an error that equality is not implemented. (Subclasses and comparisons are generally a messy business.)

It does look like you've provided proper overrides everywhere.

Reply:
Is it common to make `Base.__eq__()`  raise `NotImplementedError`, move the logic into a helper, `Base.eq`, for example, and have `SubClass.__eq__` call `super().eq`, instead of `super.__eq__()` ?